Systimer using platform tick timer

kuggenhoffen opened the pull request:
@kjbracey-arm

TeroJaasko commented:
test this please

TeroJaasko commented on bare-metal/system_timer.c:
```diff
@@ -76,38 +119,29 @@ void timer_sys_init(void)
-    }
-    return ret_val;
+    platform_tick_timer_start(TIMER_SYS_TICK_PERIOD);
 }
```
Isn't the return value missing?

kuggenhoffen replied:
yes, fixing

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
-static void timer_sys_interrupt(int8_t timer_id, uint16_t slots);
+static void timer_sys_interrupt(void);
+
+#ifndef EVENTLOOP_USE_TICK_TIMER
```
I guess this should be NS_EVENTLOOP_USE_TICK_TIMER, to namespace it a little.

kuggenhoffen replied:
true

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
     ns_list_link_t link;
 } sys_timer_struct_s;
 
+#define TIMER_SLOTS_PER_MS			20
```
Formatting - tabs

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
+#define TIMER_SLOTS_PER_MS			20
 #define TIMER_SYS_TICK_PERIOD       10 // milliseconds
-#define TIMER_SYS_TICK_SLOTS        (TIMER_SYS_TICK_PERIOD * 20) // 50us slots
+#define TIMER_SYS_TICK_SLOTS        (TIMER_SYS_TICK_PERIOD * TIMER_SLOTS_PER_SECOND) // 50us slots
```
SLOTS_PER_MS, surely. (Is this not used?)

kuggenhoffen replied:
Yes should be SLOTS_PER_MS, though I guess this is not used now since the systimer period is given as milliseconds when starting it and that is converted to slots using TIMER_SLOTS_PER_MS.

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
+
+#ifndef EVENTLOOP_USE_TICK_TIMER
+/* Implement platform tick timer using eventOS timer */
+int8_t platform_tick_timer_start(uint32_t milliseconds);
```
This declaration shouldn't be here

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
+    // Not interested in timer id or slots
+    (void)slots;
+    // Call the tick timer callback
+    if (NULL != tick_timer_callback && timer_id == tick_timer_id) {
```
Not a fan of backwards comparisons, but I'll let it pass if the rest of the file is doing it.

kuggenhoffen replied:
I'll switch this around

kjbracey-arm commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +40,59 @@ typedef struct sys_timer_struct_s {
+    (void)slots;
+    // Call the tick timer callback
+    if (NULL != tick_timer_callback && timer_id == tick_timer_id) {
+        platform_tick_timer_start(TIMER_SYS_TICK_PERIOD);
```
There's a subtle behaviour change here - you've reversed the order of the retrigger and the callback.

I'm not sure if this is good or bad. It's good in that it will reduce drift, and existing platforms more like the self-retriggering platforms-with-tickers.

But it's potentially bad in that it will change timing of existing platforms. High load saturating the event queue could conceivably cause issues.

Suggest we try it this way and see if we see anything bad...

kuggenhoffen replied:
I was debating this a bit as well and didn't notice that this might cause issue with existing platforms. Not really feeling like introducing any potential regression but I guess we can do this if we keep an eye out for possible issues...

kjbracey-arm commented on nanostack-event-loop/platform/arm_hal_timer.h:
```diff
@@ -48,6 +48,31 @@ extern void platform_timer_disable(void);
  * \return 50us time slot remaining
  */
 extern uint16_t platform_timer_get_remaining_slots(void);
+/**
```
Maybe bracket this section in the ifdef? Needs to be clear this only needs to be provided by the platform if that define is active.

kjbracey-arm commented on nanostack-event-loop/platform/arm_hal_timer.h:
```diff
@@ -48,6 +48,31 @@ extern void platform_timer_disable(void);
+/**
+ * \brief This function is API for starting the low resolution tick timer
+ *
+ * \param milliseconds define how many milliseconds time period will be started
```
"-1 for failure". Be explicit that this gets called periodically until stop is called.

kjbracey-arm commented:
Nearly there. But this is also the time to unify the two directories - source and bare-metal.

The MINAR mbed OS 3 version can now be implemented via the platform_ticker API, leaving just the "bare-metal" version here, renamed back to "source".

kuggenhoffen commented:
@kjbracey-arm @TeroJaasko 
I addressed most of your comments. Also added configuration parameter to exclude the highres timer from build. Comments on these changes welcome.

I'll start working on combining the source and bare-metal directories next.

kuggenhoffen commented on bare-metal/system_timer.c:
```diff
@@ -35,19 +36,54 @@ typedef struct sys_timer_struct_s {
+static void timer_sys_interrupt(void);
 
+#ifndef NS_EVENTLOOP_USE_TICK_TIMER
+int8_t platform_tick_timer_start(uint32_t milliseconds);
```
@kjbracey-arm I think this has to be here because in this case this is not declared in the header. Or if you prefer moving the implementation before the callback function?

kjbracey-arm replied:
Ah, good point. If not NS_EVENTLOOP_USE_TICKER_TIMER, then either it should be documented that eventloop provides the platform ticker functions ourselves, and document it as such, or they should be private and internal, meaning static definitions.

TeroJaasko commented:
Before merging, please fix the source file locations in Makefile also so the nanosimulator will compile again.

kuggenhoffen commented:
@kjbracey-arm Can you check if this is looking ok now?

kjbracey-arm commented on nanostack-event-loop/platform/arm_hal_timer.h:
```diff
@@ -15,6 +15,11 @@
 #ifndef ARM_HAL_TIMER_H_
 #define ARM_HAL_TIMER_H_
+
+#ifdef MBED_CONF_NANOSTACK_EVENTLOOP_USE_PLATFORM_TICK_TIMER
```
Think this should be "#if" - would activate if they set it to "false".

kjbracey-arm commented on nanostack-event-loop/platform/arm_hal_timer.h:
```diff
@@ -48,6 +53,33 @@ extern void platform_timer_disable(void);
+ * \return -1 for failure, success otherwise
+ */
+extern int8_t platform_tick_timer_register(void (*tick_timer_cb_handler)(void));
+/**
```
Blank line between function and next comment

kjbracey-arm commented on source/event.c:
```diff
@@ -228,7 +228,9 @@ void eventOS_scheduler_init(void)
     }
 
     /* Init Generic timer module */
+#ifndef NS_EXCLUDE_HIGHRES_TIMER
```
Hmm, that's a little awkward. You don't want to call this, and thus call the platform_timer code, if you're in tick mode.

Can you change ns_timer and thus take this out? Make it register the callback on the first actual call to register a timer?

kjbracey-arm commented on source/minar_hal_timer.cpp:
```diff
@@ -0,0 +1,70 @@
+#include "platform/arm_hal_timer.h"
+#include "platform/arm_hal_interrupt.h"
+
+#if defined(NS_EVENTLOOP_USE_TICK_TIMER) && defined(YOTTA_CFG)
```
I think "NS_EVENTLOOP_USE_TICK_TIMER" could be implicit for mbed OS 3.

You could do

    #ífdef TARGET_LIKE_MBED && defined YOTTA_CFG
    #define NS_EVENTLOOP_USE_TICK_TIMER
    #endif

in the header file.

Or can you get NS_EVENTLOOP_USE_TICK_TIMER defined in a Yotta JSON file?

kjbracey-arm commented on source/system_timer.c:
```diff
@@ -35,19 +36,54 @@ typedef struct sys_timer_struct_s {
+#define TIMER_SLOTS_PER_MS          20
 #define TIMER_SYS_TICK_PERIOD       10 // milliseconds
-#define TIMER_SYS_TICK_SLOTS        (TIMER_SYS_TICK_PERIOD * 20) // 50us slots
+#define TIMER_SYS_TICK_SLOTS        (TIMER_SYS_TICK_PERIOD * TIMER_SLOTS_PER_SECOND) // 50us slots
```
PER_MS, surely? Is this define used? (Think I said this before)

kjbracey-arm commented on source/timer_sys.h:
```diff
@@ -16,10 +16,6 @@
-extern "C" {
-#endif
-
 /**
```
Don't you need the cplusplus markings? Or is this just internal to C code now?

kuggenhoffen replied:
These got removed accidentally, not sure if these are needed now so I will add these back.

kjbracey-arm commented:
Looking pretty good.

kuggenhoffen commented:
test this please

kuggenhoffen commented:
@kjbracey-arm Previous feedback should be fixed now

kjbracey-arm commented on nanostack-event-loop/platform/arm_hal_timer.h:
```diff
@@ -16,7 +16,7 @@
 
-#ifdef MBED_CONF_NANOSTACK_EVENTLOOP_USE_PLATFORM_TICK_TIMER
+#if MBED_CONF_NANOSTACK_EVENTLOOP_USE_PLATFORM_TICK_TIMER
 #define NS_EVENTLOOP_USE_TICK_TIMER MBED_CONF_NANOSTACK_EVENTLOOP_USE_PLATFORM_TICK_TIMER
```
Normally internally we use ifdef, so I had in mind

#if MBED_CONF_NANOSTACK_EVENTLOOP_USE_PLATFORM_TICK_TIMER
#define NS_EVENTLOOP_USE_TICK_TIMER 1
#endif

And then use ifdef with NS_EVENTLOOP_USE_TICKER_TIMER

kjbracey-arm commented:
Fine with this, just do the ifdef tweak then go.

kuggenhoffen commented:
@SeppoTakalo @tommikas 
This is now ready to be merged, you guys wanna do something about those failing yotta builds or ok if I merge this regardless of those?
